File: apps/kotlin/src/test/kotlin/chapter10/EvaluateTest.kt
package chapter10

import chapter02.TrainTestSplit
import org.jetbrains.kotlinx.dataframe.AnyFrame
import org.jetbrains.kotlinx.dataframe.api.dataFrameOf
import kotlin.test.Test
import kotlin.test.assertEquals

private class AlwaysSetosa : Classifier {
    override fun fit(
        x: AnyFrame,
        t: List<String>,
    ): AlwaysSetosa = this

    override fun predict(x: AnyFrame): List<String> = List(x.rowsCount()) { "setosa" }
}

private fun smallSplit(): TrainTestSplit<String> =
    TrainTestSplit(
        xTrain = dataFrameOf("花弁幅" to listOf(0.1, 0.2, 0.8, 0.9)),
        xTest = dataFrameOf("花弁幅" to listOf(0.15, 0.25)),
        tTrain = listOf("setosa", "setosa", "virginica", "virginica"),
        tTest = listOf("setosa", "setosa"),
    )

class EvaluateTest {
    @Test
    fun `学習させてから訓練データとテストデータの正解率を求める`() {
        assertEquals(Score(train = 0.5, test = 1.0), evaluate(AlwaysSetosa(), smallSplit()))
    }

    @Test
    fun `第3章の決定木と自作のモデルを同じ関数で評価できる`() {
        val models: List<Classifier> =
            listOf(
                DecisionTreeClassifier(maxDepth = 1),
                LogisticRegression(),
                RandomForest(nEstimators = 5, maxFeatures = 1, seed = 0),
            )

        val scores = models.map { evaluate(it, smallSplit()) }

        assertEquals(List(3) { Score(train = 1.0, test = 1.0) }, scores)
    }
}

File: apps/node/test/chapter10/evaluate.test.ts
import { describe, expect, it } from "vitest";
import type { TrainTestSplit } from "../../src/chapter02/iris-preprocessing.ts";
import { DecisionTree } from "../../src/chapter03/decision-tree.ts";
import { type Classifier, evaluate } from "../../src/chapter10/classifier.ts";
import { LogisticRegression } from "../../src/chapter10/logistic-regression.ts";
import { RandomForest } from "../../src/chapter10/random-forest.ts";

type PetalWidth = Record<"花弁幅", number>;

class AlwaysSetosa implements Classifier<"花弁幅"> {
  fit(): this {
    return this;
  }

  predict(x: readonly PetalWidth[]): string[] {
    return x.map(() => "setosa");
  }
}

function smallSplit(): TrainTestSplit<PetalWidth, string> {
  return {
    xTrain: [
      { 花弁幅: 0.1 },
      { 花弁幅: 0.2 },
      { 花弁幅: 0.8 },
      { 花弁幅: 0.9 },
    ],
    xTest: [{ 花弁幅: 0.15 }, { 花弁幅: 0.25 }],
    tTrain: ["setosa", "setosa", "virginica", "virginica"],
    tTest: ["setosa", "setosa"],
  };
}

describe("evaluate", () => {
  it("学習させてから訓練データとテストデータの正解率を求める", () => {
    expect(evaluate(new AlwaysSetosa(), smallSplit())).toEqual({
      train: 0.5,
      test: 1,
    });
  });

  it("第 3 章の決定木と自作のモデルを同じ関数で評価できる", () => {
    const models: Classifier<"花弁幅">[] = [
      new DecisionTree({ maxDepth: 1 }),
      new LogisticRegression(),
      new RandomForest({ nEstimators: 5, maxFeatures: 1, seed: 0 }),
    ];

    const scores = models.map((model) => evaluate(model, smallSplit()));

    expect(scores).toEqual([
      { train: 1, test: 1 },
      { train: 1, test: 1 },
      { train: 1, test: 1 },
    ]);
  });
});
